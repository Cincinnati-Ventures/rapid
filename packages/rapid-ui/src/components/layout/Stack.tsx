import React, { useMemo } from 'react';
import { RapidStyles, getValidReactChildren } from '../../utils';

type Spacing = 'sm' | 'md' | 'lg';
type Direction = 'row' | 'column';

const RAPID_CLASSNAME = 'rapid-stack';

interface StackProps
	extends React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLDivElement>,
		HTMLDivElement
	> {
	spacing?: Spacing;
	direction?: Direction;
	styles?: string;
}


// ------------------- TAILWINDCSS CLASSES TO FORCE JIT TO COMPILE -------------------
// space-y-6
// space-y-3
// space-y-1.5
// space-x-6
// space-x-3
// space-x-1.5

const Stack = React.forwardRef<HTMLDivElement, StackProps>(
	(
		{ styles, children, spacing = 'lg', direction = 'column', ...rest },
		ref,
	) => {
		// We only want to re-run this if children change
		const sanitizedChildren = useMemo(
			() => getValidReactChildren(children),
			[children],
		);

		const getSpacing = (spacing: Spacing) => {
			switch (spacing) {
				case 'sm':
					return '1.5';
				case 'md':
					return '3';
				case 'lg':
					return '6';
			}
		};

		const getDirectionStyles = (direction: Direction) => {
			switch (direction) {
				case 'row':
					return `flex space-x-${getSpacing(spacing)}`;
				case 'column':
					return `flex flex-col space-y-${getSpacing(spacing)}`;
			}
		};
		// This component does not support custom dividers or wrapping (maybe a TODO ?)
		return (
			<div
				ref={ref}
				{...rest}
				className={RapidStyles(
					styles || rest.className,
					getDirectionStyles(direction),
					RAPID_CLASSNAME
				)}
			>
				{sanitizedChildren}
			</div>
		);
	},
);

export default Stack;
