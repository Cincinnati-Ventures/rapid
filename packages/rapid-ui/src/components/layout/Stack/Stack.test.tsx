import React from 'react';
import { render } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import Stack from './Stack';

describe('Stack', () => {
	it('should render default correctly', () => {
		const { container } = render(<Stack />);
		expect(container).toMatchSnapshot();
	});

	it('should render correctly with non default props', () => {
		const { container } = render(
			<Stack styles='bg-blue-500' spacing='md' direction='row' />,
		);
		expect(container).toMatchSnapshot();
	});

	it('should render correctly with children', () => {
		const { container } = render(
			<Stack styles='bg-red-500' spacing='md'>
				<div>
					<p>Hello world!</p>
				</div>
				<div>
					<p>Hello world!</p>
				</div>
				<div>
					<p>Hello world!</p>
				</div>
				<div>
					<p>Hello world!</p>
				</div>
			</Stack>,
		);
		expect(container).toMatchSnapshot();
	});
});
