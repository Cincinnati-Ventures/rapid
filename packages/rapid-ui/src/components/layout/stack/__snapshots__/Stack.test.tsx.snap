// Vitest Snapshot v1

exports[`Stack > should render correctly with children 1`] = `
<div>
  <div
    class="flex flex-col space-y-3 bg-red-500 rapid-stack"
  >
    <div>
      <p>
        Hello world!
      </p>
    </div>
    <div>
      <p>
        Hello world!
      </p>
    </div>
    <div>
      <p>
        Hello world!
      </p>
    </div>
    <div>
      <p>
        Hello world!
      </p>
    </div>
  </div>
</div>
`;

exports[`Stack > should render correctly with non default props 1`] = `
<div>
  <div
    class="flex space-x-3 bg-blue-500 rapid-stack"
  />
</div>
`;

exports[`Stack > should render default correctly 1`] = `
<div>
  <div
    class="flex flex-col space-y-6 rapid-stack"
  />
</div>
`;
